### PR TITLE
Implement club fair pages

### PR DESCRIPTION
**Pull request by ezwang**
**Prerequisites**
- Merge after #59 
- Merge after https://github.com/pennlabs/clubs-backend/pull/25, https://github.com/pennlabs/clubs-backend/pull/31, and https://github.com/pennlabs/clubs-backend/pull/34

**Features**
- Implements flyer page(s) with QR code
- Implements club fair page
- Implements school/major/graduation year profile page
- Implements first time login welcome page
- Implements subscriber download page

**Screenshots**
- image
- image
- (old) image
- image
- image
- image
- image
- (old) image

### CODE REVIEW

File: components/ClubPage/InfoBox.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ const InfoBox = props => {
   ]
 
   return data.map(({ icon, text }) => (
-    <p style={infoStyles}>
+    <p style={infoStyles} key={text}>
```
Comment (cacabo) [inline components/ClubPage/InfoBox.js]: Can we use the shared `Text` component from the shared components?

File: components/common/Icon.js [hunk 2]
```diff
@@ -19,6 +19,15 @@ const IconTag = s.img`
     margin-right: 3px;
     margin-top: -5px;
   }
+
+  .button.is-large & {
```
Comment (cacabo) [inline components/common/Icon.js]: Why is there an & after `.is-large`?

File: pages/club/[club]/fair.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          <TagGroup tags={club.tags} />
+        </Margin>
+        <Text>{club.description}</Text>
+        <Link href="/club/[club]" as={`/club/${club.code}`}>
+          <a className="button is-danger is-large">See more details</a>
+        </Link>
+      </Center>
+    </WideContainer>
+  )
+}
```
Comment (cacabo) [inline pages/club/[club]/fair.js]: Great use of the shared components :)

File: pages/club/[club]/index.js [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-import Tabs from '../components/ClubPage/Tabs'
-import Header from '../components/ClubPage/Header'
-import InfoBox from '../components/ClubPage/InfoBox'
-import SocialIcons from '../components/ClubPage/SocialIcons'
+import renderPage from '../../../renderPage.js'
+import { doApiRequest } from '../../../utils'
+import Tabs from '../../../components/ClubPage/Tabs'
+import Header from '../../../components/ClubPage/Header'
+import InfoBox from '../../../components/ClubPage/InfoBox'
+import SocialIcons from '../../../components/ClubPage/SocialIcons'
```
Comment (cacabo) [inline pages/club/[club]/index.js]: Ye this is why module aliases will be great

File: pages/invite/[club]/[invite]/[token]/index.js [hunk 5]
```diff
@@ -1,7 +1,7 @@
-import renderPage from '../renderPage.js'
-import { doApiRequest, formatResponse, LOGIN_URL } from '../utils'
+import renderPage from '../../../../../renderPage.js'
+import { doApiRequest, formatResponse, LOGIN_URL } from '../../../../../utils.js'
```
Comment (cacabo) [inline pages/invite/[club]/[invite]/[token]/index.js]: Why have `.js` on the end of these?

File: pages/club/[club]/index.js [hunk 6]
```diff
@@ -1,12 +1,12 @@
 import { useState, useEffect } from 'react'
 import s from 'styled-components'
 
-import renderPage from '../renderPage.js'
-import { doApiRequest } from '../utils'
-import Tabs from '../components/ClubPage/Tabs'
-import Header from '../components/ClubPage/Header'
-import InfoBox from '../components/ClubPage/InfoBox'
-import SocialIcons from '../components/ClubPage/SocialIcons'
+import renderPage from '../../../renderPage.js'
```
Comment (cacabo) [inline pages/club/[club]/index.js]: Why have `.js` on the end of this?

File: components/ClubPage/InfoBox.js (hunk 1 above)
Comment (ezwang) [reply components/ClubPage/InfoBox.js]: I don't think this was introduced by me, but instead upstream in #59. Can fix in this merge request, just note that the same issue will appear in #59.

File: components/common/Icon.js (hunk 2 above)
Comment (ezwang) [reply components/common/Icon.js]: We want to apply these rules to the icon, not the parent button. The icon is nested inside the parent button, it is not the button.

File: pages/invite/[club]/[invite]/[token]/index.js (hunk 5 above)
Comment (ezwang) [reply pages/invite/[club]/[invite]/[token]/index.js]: Fixed, note that these changes were introduced in #59 and will also appear in that merge request.

File: pages/club/[club]/index.js (hunk 6 above)
Comment (ezwang) [reply pages/club/[club]/index.js]: Fixed, note that these changes were introduced in #59 and will also appear in that merge request.

File: utils.js [hunk 7]
```diff
@@ -74,7 +78,7 @@ export function doApiRequest(path, data) {
   if (data.body && !(data.body instanceof FormData)) {
     data.body = JSON.stringify(data.body)
   }
-  return fetch(API_BASE_URL + path, data)
+  return fetch(getApiUrl(path), data)
```
[the remark this answers is not in the document]
Comment (ezwang) [reply utils.js]: I use this function later on where I don't actually need to do an API request, but instead obtain a URL where the user can download an Excel spreadsheet.

Ideally I would want them all to use the same function, that way if we need to update the backend URL we only change it in one place.

We could also use use `API_BASE_URL + string` everywhere, but I'm not sure if that's cleaner.

File: pages/club/[club]/edit.js [hunk 8]
```diff
@@ -284,6 +285,13 @@ class ClubForm extends React.Component {
...
           })
         )
+      doApiRequest(`/clubs/${clubId}/subscription/?format=json`)
+        .then(resp => resp.json())
+        .then(data =>
+          this.setState({
+            subscriptions: data,
+          })
+        )
     }
```
[the remark this answers is not in the document]
Comment (ezwang) [reply pages/club/[club]/edit.js]: This depends on the user being logged in, we don't want non-officers/owners or unauthenticated users to be able to see the list of people's emails and full names.

File: pages/club/[club]/fair.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  }
+`
+
+const Fair = ({ authenticated, query, userInfo, favorites, updateFavorites, subscriptions, updateSubscriptions }) => {
+  const [club, setClub] = useState(null)
+
+  useEffect(() => {
+    doApiRequest(`/clubs/${query.club}/?format=json`)
+      .then(resp => resp.json())
+      .then(data => setClub(data))
```
[the remark this answers is not in the document]
Comment (ezwang) [reply pages/club/[club]/fair.js]: This request can be done in `getInitialProps`, I'll work on converting it over.

File: utils.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (ezwang) [reply utils.js]: What do you think? Should we use the function or switch to `API_BASE_URL + string`?

File: pages/club/[club]/fair.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ezwang) [reply pages/club/[club]/fair.js]: Should we do this once #56 gets resolved and merged in? Perhaps include this change in that PR too. There seems to be some discussion on error handling that needs to be resolved, and that would apply to this case too.

File: utils.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (sigmachirality) [reply utils.js]: I defer to @cacabo

File: pages/club/[club]/fair.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (sigmachirality) [reply pages/club/[club]/fair.js]: if that PR gets merged in I think a lot of pages could eventually be converted to SSR or get optimized as static pages, so we can audit/discuss this later

[the remark this answers is not in the document]
Comment (ezwang) [reply pages/club/[club]/fair.js]: Great, I'll mark as resolved for now.

File: utils.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (cacabo) [reply utils.js]: I think this works fine for now. Have plans for a slight cleanup / abstraction of API requests to decouple frontend logic from  API-interfacing logic but this works well enough.

[the remark this answers is not in the document]
Comment (ezwang) [reply utils.js]: I agree, API logic could definitely be cleaned up. We can do so in a later PR.

**Comment by ezwang** [comment]
This PR can be merged as soon as https://github.com/pennlabs/clubs-backend/pull/25 gets merged!

There's just one minor feature (a link to a download) that won't work until that PR is merged in. This could even be merged in beforehand and we can leave that link non-functional until the PR gets merged.

**Pull request merged**